### PR TITLE
Added 'dump' command for partitions

### PR DESCRIPTION
**Pull request by andy-maier**
See the commit message for details.
This PR requires the latest master version of the zhmcclient project to be installed (e.g. `pip install .` or `pip install -e .`)

Ready for review and trying it out. To display help: `zhmc partition dump --help`

TODO:
- Pull in zhmcclient 0.29.0 once released.

### CODE REVIEW

File: zhmccli/_cmd_partition.py [hunk 1]
```diff
@@ -146,6 +145,58 @@ def partition_stop(cmd_ctx, cpc, partition):
...
+              'on the storage array (also shown in the HMC GUI). The storage '
+              'volume may be of type FCP or FICON. '
+              'For CPCs without the storage management feature (z13 and '
+              'earlier): A string of the form "HBA.WWPN.LUN", where HBA is the '
+              'name of the HBA resource in the partition and WWPN and LUN '
+              'identify the storage array and storage volume thereon. The '
+              'storage volume must be of type FCP.')
+@click.option('--lba', type=str, required=False, default='0',
+              help='The logical block number of the anchor point for locating '
+              'the dump program on the storage volume. Default: 0')
```
Comment (mhartmay) [inline zhmccli/_cmd_partition.py:167]: General question: Why do we duplicate the default values? Why do we not use the option `show_default=True` instead https://click.palletsprojects.com/en/7.x/api/#click.Option (maybe we can change this in a separate patch)?

File: zhmccli/_cmd_partition.py [hunk 2]
```diff
@@ -28,11 +28,10 @@
 from .zhmccli import cli, CONSOLE_LOGGER_NAME
 from ._helper import print_properties, print_resources, abort_if_false, \
     options_to_properties, original_options, COMMAND_OPTIONS_METAVAR, \
-    part_console, raise_click_exception
+    part_console, raise_click_exception, storage_management_feature
 from ._cmd_cpc import find_cpc
 from ._cmd_metrics import get_metric_values
 
-
```
Comment (mhartmay) [inline zhmccli/_cmd_partition.py:35]: Seems unrelated?

File: zhmccli/_cmd_partition.py [hunk 3]
```diff
@@ -916,6 +967,133 @@ def cmd_partition_console(cmd_ctx, cpc_name, partition_name, options):
...
+        'volume': None,
+        'timeout': None,
+    }
+    options = original_options(options)
+    properties = options_to_properties(options, name_map)
+    volume_opt = options['volume']  # It is required
+    timeout = options['timeout']  # It is optional but has a default
+
+    if storage_management_feature(partition):
+
```
Comment (mhartmay) [inline zhmccli/_cmd_partition.py:994]: Is this blank line inteded?

File: zhmccli/_cmd_partition.py [hunk 4]
```diff
@@ -916,6 +967,133 @@ def cmd_partition_console(cmd_ctx, cpc_name, partition_name, options):
...
+        # Check and parse the --volume option into SG.SV or UUID
+        volume_parts = volume_opt.split('.')
+        if len(volume_parts) == 1:  # format: UUID
+            uuid = volume_parts[0]
```
Comment (mhartmay) [inline zhmccli/_cmd_partition.py:998]:
```suggestion
            uuid, = volume_parts
```

File: zhmccli/_cmd_partition.py [hunk 5]
```diff
@@ -916,6 +967,133 @@ def cmd_partition_console(cmd_ctx, cpc_name, partition_name, options):
...
+    options = original_options(options)
+    properties = options_to_properties(options, name_map)
+    volume_opt = options['volume']  # It is required
+    timeout = options['timeout']  # It is optional but has a default
+
+    if storage_management_feature(partition):
+
+        # Check and parse the --volume option into SG.SV or UUID
+        volume_parts = volume_opt.split('.')
+        if len(volume_parts) == 1:  # format: UUID
```
Comment (mhartmay) [inline zhmccli/_cmd_partition.py:997]: It's quite obvious what we're doing here... but something like
```if is_uuid(volume_parts):``` wouldn't require even a comment.

File: zhmccli/_cmd_partition.py [hunk 6]
```diff
@@ -916,6 +967,133 @@ def cmd_partition_console(cmd_ctx, cpc_name, partition_name, options):
...
+                except zhmcclient.NotFound:
+                    continue
+            if not storage_volume:
+                raise_click_exception(
+                    "Storage volume with UUID '{uuid}' specified in the "
+                    "--volume option was not found in any storage group "
+                    "attached to partition '{part}' of CPC '{cpc}'.".
+                    format(uuid=uuid, part=partition_name, cpc=cpc_name),
+                    cmd_ctx.error_format)
+        elif len(volume_parts) == 2:  # format: SG.SV
```
Comment (mhartmay) [inline zhmccli/_cmd_partition.py:1014]: Maybe `is_...(volume_opt)`.

File: zhmccli/_cmd_partition.py [hunk 7]
```diff
@@ -916,6 +967,133 @@ def cmd_partition_console(cmd_ctx, cpc_name, partition_name, options):
...
+                if _sg.name == sg_name:
+                    sg = _sg
+                    break
+            if not sg:
+                raise_click_exception(
+                    "Storage group '{sg}' specified in the --volume option was "
+                    "not found in the storage groups attached to partition "
+                    "'{part}' of CPC '{cpc}'.".
+                    format(sg=sg_name, part=partition_name, cpc=cpc_name),
+                    cmd_ctx.error_format)
```
Comment (mhartmay) [inline zhmccli/_cmd_partition.py:1028]: Maybe a little bit overkill...
```suggestion
try:
  sg = next(filter(lambda sg_ele: sg_ele.name == sg_name, sg_list))
except StopIteration:
  raise ...
```

File: zhmccli/_cmd_partition.py [hunk 8]
```diff
@@ -916,6 +967,133 @@ def cmd_partition_console(cmd_ctx, cpc_name, partition_name, options):
...
+        properties['storage-volume-uri'] = storage_volume.uri
+        properties['timeout'] = timeout
+
+        op_properties = {
+            'dump-program-info': properties,
+            'dump-program-type': 'storage',
+        }
+        try:
+            partition.start_dump_program(
+                parameters=op_properties, wait_for_completion=True)
```
Comment (mhartmay) [inline zhmccli/_cmd_partition.py:1057]: Is there no timeout support for start_dump_program?

File: zhmccli/_cmd_partition.py [hunk 9]
```diff
@@ -916,6 +967,133 @@ def cmd_partition_console(cmd_ctx, cpc_name, partition_name, options):
...
+        except zhmcclient.Error as exc:
+            raise_click_exception(exc, cmd_ctx.error_format)
+
+    else:  # No storage management feature
+
+        # Check and parse the --volume option into HBA.WWPN.LUN
+        volume_parts = volume_opt.split('.')
```
Comment (mhartmay) [inline zhmccli/_cmd_partition.py:1064]: Maybe introduce a function for parsing a HBA.WWPN.LUN string?

File: zhmccli/_cmd_partition.py [hunk 10]
```diff
@@ -916,6 +967,133 @@ def cmd_partition_console(cmd_ctx, cpc_name, partition_name, options):
...
+            hba = partition.hbas.find(name=hba_name)
+        except zhmcclient.NotFound:
+            raise_click_exception(
+                "HBA '{hba}' specified in the --volume option was not found "
+                "in partition '{part}' of CPC '{cpc}'.".
+                format(hba=hba_name, part=partition_name, cpc=cpc_name),
+                cmd_ctx.error_format)
+
+        # Set the remaining feature-specific parameters
+        properties['dump-load-hba-uri'] = hba.uri
```
Comment (mhartmay) [inline zhmccli/_cmd_partition.py:1083]: Is it okay that we reuse the properties dict here?

File: zhmccli/_cmd_partition.py [hunk 11]
```diff
@@ -916,6 +967,133 @@ def cmd_partition_console(cmd_ctx, cpc_name, partition_name, options):
...
+            'dump-program-type': 'storage',
+        }
+        try:
+            partition.start_dump_program(
+                parameters=op_properties, wait_for_completion=True)
+        except zhmcclient.Error as exc:
+            raise_click_exception(exc, cmd_ctx.error_format)
+
+    else:  # No storage management feature
+
```
Comment (mhartmay) [inline zhmccli/_cmd_partition.py:1062]: I would probably factor out the with-storage-management-dump and the without-storage-management-dump code into two separate functions, otherwise this function is pretty long and harder to test.

```
if storage_management_feature(partition):
    dump_....(...)
else:
    dump_....(...)
```

File: zhmccli/_cmd_partition.py [hunk 12]
```diff
@@ -916,6 +967,133 @@ def cmd_partition_console(cmd_ctx, cpc_name, partition_name, options):
...
+    }
+    options = original_options(options)
+    properties = options_to_properties(options, name_map)
+    volume_opt = options['volume']  # It is required
+    timeout = options['timeout']  # It is optional but has a default
+
+    if storage_management_feature(partition):
+
+        # Check and parse the --volume option into SG.SV or UUID
+        volume_parts = volume_opt.split('.')
```
Comment (mhartmay) [inline zhmccli/_cmd_partition.py:996]: This doesn't work since `.` can be used in a volume name and storage name.
The allowed pattern for volume and storage name is `[a-zA-Z0-9_.@- ]{1,64}` (please double check).

We can either use a separate cmdline option or, for example, `/` as delimiter.

File: zhmccli/_cmd_partition.py [hunk 13]
```diff
@@ -916,6 +967,133 @@ def cmd_partition_console(cmd_ctx, cpc_name, partition_name, options):
...
+            raise_click_exception(
+                "Invalid format for --volume option: '{v}'. CPC '{cpc}' has "
+                "the storage management feature which requires the formats "
+                "'SG.SV' or 'UUID'.".
+                format(v=volume_opt, cpc=cpc_name),
+                cmd_ctx.error_format)
+
+        # Set the remaining feature-specific parameters
+        properties['storage-volume-uri'] = storage_volume.uri
+        properties['timeout'] = timeout
```
Comment (mhartmay) [inline zhmccli/_cmd_partition.py:1049]: I got the following error when this line was not commented...
`Error: HTTPError: 400,14: Invalid value for query parameter timeout was specified.  [POST /api/partitions/.../operations/start-dump-program]`

File: zhmccli/_cmd_partition.py [hunk 14]
```diff
@@ -916,6 +967,133 @@ def cmd_partition_console(cmd_ctx, cpc_name, partition_name, options):
...
+
+        op_properties = {
+            'dump-program-info': properties,
+            'dump-program-type': 'storage',
+        }
+        try:
+            partition.start_dump_program(
```
Comment (mhartmay) [inline zhmccli/_cmd_partition.py:1056]: Is it intended that no timeout is specified?

File: zhmccli/_cmd_partition.py (hunk 1 above)
Comment (andy-maier) [reply zhmccli/_cmd_partition.py:167]: This would definitely be a separate issue/PR because it would affect all commands. I have opened issue #125.
RESOLVED

File: zhmccli/_cmd_partition.py (hunk 2 above)
Comment (andy-maier) [reply zhmccli/_cmd_partition.py:35]: DONE - Changed back.

File: zhmccli/_cmd_partition.py (hunk 3 above)
Comment (andy-maier) [reply zhmccli/_cmd_partition.py:994]: Yes, for better readability because the next line comment belongs to the code after it.

File: zhmccli/_cmd_partition.py (hunk 12 above)
Comment (andy-maier) [reply zhmccli/_cmd_partition.py:996]: Good point. I have double checked the names for SG and SV and the pattern you show is correct.
DONE - changed to using `/` as a delimiter, so it is now `SG/SV`.

For consistency, I also changed the delimiter used for the string without storage mgmt feature from dot to slash, so it is now `HBA/WWPN/LUN`.

File: zhmccli/_cmd_partition.py (hunk 5 above)
Comment (andy-maier) [reply zhmccli/_cmd_partition.py:997]: I would not like to introduce an `is_uuid()` function because the format parsing would then be done inside that function and again on the caller's side to get the values.
The comment documents what conclusion is drawn from the if condition, since the condition itself does not make that obvious (which lead you to propose the function).
REJECTED - No change.

File: zhmccli/_cmd_partition.py (hunk 4 above)
Comment (andy-maier) [reply zhmccli/_cmd_partition.py:998]: I don't understand the motivation for this change. Besides that, the new code does not work on Python 2.7:
```
Python 2.7.16 (default, Sep  2 2019, 11:59:44) 
[GCC 4.2.1 Compatible Apple LLVM 10.0.1 (clang-1001.0.46.4)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> a=(1,2,3)
>>> b, = a
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: too many values to unpack
>>> 
```
REJECTED.

File: zhmccli/_cmd_partition.py (hunk 6 above)
Comment (andy-maier) [reply zhmccli/_cmd_partition.py:1014]: See response above.

File: zhmccli/_cmd_partition.py (hunk 7 above)
Comment (andy-maier) [reply zhmccli/_cmd_partition.py:1028]: This changes 6 easy to understand lines into 3 lines one of which is much harder to understand.

At the end of the day, it is a matter of whether or not one likes using `filter()`. Shorter code and seemingly more elegant, but requires me every time to think for a minute to understand what it does. I can grasp what the original code does, within seconds. Python introduced list comprehensions because users found it hard to understand combinations of filter/map. So I'm not the only one in this camp it seems ;-)

I prefer using the original code.
REJECTED.

File: zhmccli/_cmd_partition.py (hunk 13 above)
Comment (andy-maier) [reply zhmccli/_cmd_partition.py:1049]: After reading through the HMC WS API documentation for Start Dump Program, it seems to me that the only way the timeout parameter can possibly be invalid is that the target volume was not a FICON volume, for which the timeout parameter is documented as invalid. Strange that it uses reason code 14 though, because that means error in the query parameters, and timeout is part of the body.

Anyway, I have changed the code to include the timeout parameter only when the target volume type (actually the type of its SG) is FICON. Please test again with the same volume type that caused the error.

File: zhmccli/_cmd_partition.py (hunk 14 above)
Comment (andy-maier) [reply zhmccli/_cmd_partition.py:1056]: I suppose you talk about the `operation_timeout` parameter that is for waiting for the asynchronous job to complete, and not about the `timeout` parameter for a FICON target volume.

To be honest, I did not think about this, but we should have a timeout capability for that.

The question is, can we combine these two timeout values to keep it simple for the users, i.e. have the `--timeout` option determine both timeout values? For now, I have changed the code to do that, but that is probably worth a discussion.

The other question is, do we need to handle the operation timing out (exception `zhmcclient.OperationTimeout`) other than currently where we end the zhmc program with an according error message. This means the asynchronous job will continue to execute, but zhmc provides no means to manage asynchronous jobs. It seems to me that we should keep not interfering with the asynchronously running job, but add a job mgmt capability to zhmc so that such jobs can be determined and managed.

**DISCUSSION**

File: zhmccli/_cmd_partition.py (hunk 11 above)
Comment (andy-maier) [reply zhmccli/_cmd_partition.py:1062]: I like that idea and have split it up. The new code also no longer needs to make the assumption about commonality of the input parameters between "Dump Partition" and "Start Dump Program".
DONE

File: zhmccli/_cmd_partition.py (hunk 9 above)
Comment (andy-maier) [reply zhmccli/_cmd_partition.py:1064]: Not worthwhile, and consistent with not doing it for the parsing of SG/SV, I would rather not do it here as well.
REJECTED.

File: zhmccli/_cmd_partition.py (hunk 10 above)
Comment (andy-maier) [reply zhmccli/_cmd_partition.py:1083]: We are not reusing the variable name for a different purpose, we continue setting it up for the same purpose as above. I have renamed it to "parameters" as part pf the splitting of the two functions.

**Comment by andy-maier** [comment]
@mhartmay Marc, I have worked through all of your comments, see above. Please review the resolutions and mark those as resolved that you agree with.

File: zhmccli/_cmd_partition.py (hunk 4 above)
Comment (mhartmay) [reply zhmccli/_cmd_partition.py]: The intention behind this was that it would make refactoring the code easier. But right, it doesn't work for Python2.

File: zhmccli/_cmd_partition.py (hunk 10 above)
Comment (mhartmay) [reply zhmccli/_cmd_partition.py]: AFAICR, the properties dict was used for two calls.

**Comment by mhartmay** [comment]
@andy-maier I tested it again on a DPM system with storage management support and it works for both: specifying the volume using UUID and SG/VG.

File: zhmccli/_cmd_partition.py (hunk 10 above)
Comment (andy-maier) [reply zhmccli/_cmd_partition.py]: Not in the current code.

**Comment by andy-maier** [comment]
Thanks for testing it and glad that it works. Merging it now.
